### DISCUSSION TITLE
.NET Core 3.1 compiles integral conversions to doubles differently than real literals. This is not true for .NET Framework.

### DISCUSSION THREAD

**Discussion by zacknewman** [dotnet/csharplang#4074] [General] [answered]
### Description

* Implicit (and explicit) casts from integrals to `double`s don't undergo the same process as real literals *during compilation*.
* Example:
```
using System;
namespace App
{
    internal static class Program
    {
        internal static void Main()
        {
            /// Displays 'True'.
           Console.WriteLine(10648738977740919977d != 10648738977740919977);
        }
    }
}
```
* [C# 6.0 draft specification](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure) states “the value of a real literal of type float or double is determined by using the IEEE ‘round to nearest’ mode” under Lexical structure -> Grammars -> Lexical grammar -> Lexical analysis -> Tokens -> Literals -> Real literals. IEEE 754 explicitly allows for extended precision format, so the compiler can compile *both* the real literal and the implicit cast using (or not) any extended precision capabilities that may exist on the CPU during compilation and still be compliant with the specification.

### Configuration

* Compiled with Microsoft Visual C# Compiler version 3.7.0-6.20459 for .NET Core 3.1.
* Windows 10 Pro 18363.1139.
* x64-based CPU with an 80-bit FPU (which has a 64-bit significand).
* .NET Framework compiles both the implicit cast and real literal the same (both *not* using the extended precision format available on my CPU).

### Other information

* As mentioned by [Eric Lippert](https://stackoverflow.com/questions/2342396/why-does-this-floating-point-calculation-give-different-results-on-different-mac/2343351#2343351), “... C# allows a higher precision type to be used for all floating-point operations”; however I was under the impression this is only true *at runtime*.

**Comment by YairHalberstadt** [comment]
It seems this is actually a difference between x64 and x86, even on .Net Framework. The compiler converts this comparison straight to a constant, but the constant is false for x86, and true for x64.

@tannergooding any ideas here?

**Reply by zacknewman** [reply]
Ah, I apologize. I carelessly assumed it was a .NET Core vs. .NET Framework issue. I didn't bother trying x86.

**Reply by YairHalberstadt** [reply]
No problem - playing around with sharplab is a great way to see what's going on here.

**Reply by zacknewman** [reply]
Can one compile under "x64" mode on an x86 CPU (only the 32-bit version)? If so, what happens? I don't own any devices that have an x86 CPU, and sharplab doesn't have an option for "x64 on x86 CPU". Even better, what happens when one compiles under "x64" mode on a CPU that *doesn't* have extended precision format capabilities?

**Reply by tannergooding** [reply]
CC. @jaredpar this is a determinism bug and looks to be caused by Roslyn depending on the runtime for the `ulong -> double` conversion. x64 vs x86 doesn't matter.

The root cause is a runtime bug in the conversion. I'll be opening a separate issue here shortly.

**Reply by zacknewman** [reply]
@tannergooding, do you want to post that as its own comment so that I can mark it as an answer?

**Comment by tannergooding** [comment, marked as answer]
This is a determinism bug and looks to be caused by Roslyn depending on the runtime for the ulong -> double conversion. x64 vs x86 doesn't matter.

The root cause is a runtime bug in the conversion code: https://github.com/dotnet/runtime/issues/43895

Roslyn could implement its own conversion algorithm for correctness such as by using the Berkeley SoftFloat 3 implementation which is known to be good/correct: https://github.com/ucb-bar/berkeley-softfloat-3/blob/master/source/ui64_to_f64.c

**Reply by zacknewman** [reply]
If it's a determinism bug by Roslyn, why is it a runtime bug and not compiler bug?

**Reply by tannergooding** [reply]
It surfaces in Roslyn as a determinism issue but is ultimately caused by its dependency on the runtime conversion implementation which is buggy.

You aren't likely to see any differences when compiling using x86/x64 but it does look to differ when compiling for ARM which has a native conversion instruction that does behave correctly.

**Reply by tannergooding** [reply]
Maybe to state it more clearly. It is a bug in both.

**Reply by zacknewman** [reply]
So what part is incorrect? The conversion of the `ulong` (which I believe is relying on extended precision format) or the conversion of the real literal (which uses 64-bit precision)? I think both conversions are *technically* correct by themselves according to IEEE 754 since it explicitly allows for extended precision format. The issue is the inconsistency.

**Reply by tannergooding** [reply]
> So what part is incorrect? 

The former, that is the issue is the conversion of a `ulong` to a `double` when the `ulong` is greater than `long.MaxValue`. There may be another bug on 32-bit with values greater than `int.MaxValue`, but I've not confirmed.

> I think both conversions are technically correct by themselves according to IEEE 754

Not quite. IEEE 754 defines several precisions but you aren't allowed to perform operations at any higher precision.
The actual spec requires that any operation take the inputs as given, perform the operation as if to infinite precision and unbounded range, and then round that result to the nearest representable value in the output format.

In order for this to be correct, the operation should ideally be done exactly in the target precision. That being said, you can also perform certain operations to at least twice the output bits and the result will not incur issues from double-rounding (provided you do ultimately round to the target precision before doing any other operations). Papers such as https://dl.acm.org/doi/10.1145/221332.221334 provide the relevant proof.

This means that doing a single 32-bit op as 64-bit and then rounding down to 32-bits is fine. Likewise doing a single 64-bit op as 128-bit. Doing 64-bit ops as 80-bits is not allowed and will introduce errors.

The 80-bit extended float exposed as part of the x87 FPU stack is also not a formally defined type in the IEEE 754 2008 or 2019 spec. It did exist abstractly as part of the 1985 spec, but it was more a "minimum" definition and not an exact one. We also largely don't use the x87 FPU stack in the current implementations and instead rely on SSE/SSE2 instructions for RyuJIT. The exception is largely on 32-bit where it is still required for returning floating-point values from methods; but where we force the rounding mode to 64-bit which largely resolves the above double rounding issues.

*Edit:* Updated a statement to clarify it must be a single op at twice the precision with rounding at each intermediate step. If you don't round after each intermediate step, error may be introduced.

**Reply by zacknewman** [reply]
That is such an amazing reply. I can't thank you enough. I'm a math guy, so I'm all about proofs. I was starting to get frustrated on Stack Overflow for the less-than-stellar and imprecise replies. Seriously, this reply made my day.

**Reply by zacknewman** [reply]
I apologize for rehashing this, but I still can’t shut up my stupid brain. Furthermore, if what I’m about to ask is stupid; then I preemptively apologize for it was well. You mention that a possible fix is for the integral conversion to be done using an algorithm intrinsic to the compiler (e.g., Berkeley SoftFloat 3), but mustn’t the compiler already be relying on some intrinsic algorithm during the lexical parsing phase when it encounters the real literal? If not, how is it avoiding the issue?

**Reply by CyrusNajmabadi** [reply]
> but mustn’t the compiler already be relying on some intrinsic algorithm during the lexical parsing phase when it encounters the real literal? If not, how is it avoiding the issue? 

The compiler is calling into the BCL method for this (the one that has hte bug).  By 'intrinsic' we mean hte compiler would ship with its own copy of the parsing algorithm that would be independent from the BCL.  It would not be the first time we've done this, for very similar reasons (i.e. isolating us from bugs in the BCL).

**Reply by tannergooding** [reply]
Right. There are two possible fixes (the compiler could take), as I see it.

The first is to reparse the integer literal directly as a `double` (the compiler already has its own floating-point parser) and the second would be to just replace the integer to double conversion with a custom implementation (it currently relies on the underlying runtime for this).
The latter is likely more future proof and less prone to break on theoretical alternative runtimes.

**Reply by Victor-Tseng** [reply]
Based on the [codegenxarch.cpp#L6435-L6468](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/codegenxarch.cpp#L6435-L6468) mentioned in [dotnet/runtime/#43895](https://github.com/dotnet/runtime/issues/43895), it looks that the code has been there for a few years? And I am able to reproduce the issue using VS2015 + 4.5.2 (older Roslyn and runtime) and VS2019 + 4.7.2 (latest Roslyn and newer runtime) on X86 CPU systems. Not sure if I just get it wrong, or the issue is not limited to .NET Core...

@zacknewman  since you are a math guy, my two cents is that you might want to consider using the decimal type instead. The decimal type has a different design and is more a WYSIWYG one. Comparing to the double's ["ten 0.1d added together doesn't equal to 0.1d*10"](https://docs.microsoft.com/en-us/dotnet/api/system.double?view=netcore-3.1#Precision) behavior, this should give you less headache.

**Reply by tannergooding** [reply]
You can also encounter similar issues with `decimal`, its just "less common" since decimal is "base 10" while `float`/`double` are "base 2".
For example, `(1m / 3m) * 3m` is `0.9999999999999999999999999999`.

Fundamentally, you will always have rounding error due to trying to store results in finite space. You even hit it with integers on overflow.
You can solve some of these problems with something like a `Rationale` type, but those then can't represent other values like `PI`, etc.

**Reply by Victor-Tseng** [reply]
@tannergooding Yes agree that `decimal `can still hit similar issues due to the finite precision.

My point is that the design of `decimal` has a consistency between the external behavior and the internal state. When you see a `decimal` variable holds the value _x_, its internal state will be _always_ exactly _x_. But for `double` variables you just don't know unless you inspect its internal state.

This consistency leads to a consistent arithmetic result, which should be critical when you are writing number-sensitive applications.

**Reply by zacknewman** [reply]
Thanks, but I am already aware of `decimal`; and as @tannergooding stated, it is still a floating point type that has rounding issues just like it does in "normal math" notation (unless you adopt the use of the repetend symbol). If I wanted to preserve accuracy, I would use something like Haskell’s `Rational`. Furthermore, the format of `decimal` is very much non-standard and doesn’t align with IEEE 754 decimal128 nor does it have the nice feature of binary64 and binary32 of having an almost identical format to how numbers are written in "normal math" (i.e., an infinite series of the base of your system raised to the index variable multiplied by coefficients in your base system).

This is largely irrelevant to the specific situation I ran into anyway. It came up when writing explicit casting ability from `U128` terms to `double` terms where `U128` is the type representing the commutative ring ℤ/2<sup>128</sup>ℤ. The algorithm I implemented follows IEEE 754 and furthermore was initially written in Rust where I was able to compare it with Rust’s own conversions since Rust has a native `u128` type. Unsurprisingly, due to the simplicity of the algorithm—due to the aforementioned property binary64 and binary32 have—I was able to confirm my implementation matched that of Rust’s for several edge cases and 1 billion randomly generated numbers—no, I did not take the time to prove my implementation was correct.

It wasn’t until I translated it to VB.NET—the language we use at my place of work—that I discovered the `ULong` term `10648738977740919977UL` was being cast differently than the equivalent `U128` term. Fortunately, this was one of the few times in life where ignorance was useful as I was *not* aware of the specification allowing floating point arithmetic to occur on CPUs with extended precision formats *including non-compliant ones that rely on the x87 FPU stack as stated by @tannergooding*. Had I known that, the discrepancy would have been explicable; and I would have stopped there.

To ensure nothing got lost in translation, I grabbed the fist byte of the `Double` term returned from my casting algorithm—I’m using a little-endian CPU—as I was pretty confident the issue lied with rounding. To my delight, it matched the value returned from Rust’s algorithm. At this point, I suspected something was “fishy” with VB.NET (and later confirmed in C# too). You can imagine my dismay when I found that `10648738977740919977R` (i.e., `10648738977740919977d` in C#) had the *same* value! How was that possible?!? I used the same machine, same compiler (even directly invoking the compiler itself with the same options), and targeted the same platform. Finally, I realized that I was not actually checking the same thing as I had used a real literal; so it wasn’t until I examined the `Double` term that resulted from an implicit cast of the `ULong` term `10648738977740919977UL` that I found the rounding discrepancy. What is funny, is that had there not been this bug, I likely would still be in a state of befuddlement since it too would have resulted in the same value. Not sure if I would have figured out that the integral value had to be generated at runtime to get the different behavior; although in hindsight, it all makes sense.

**Reply by zacknewman** [reply]
@tannergooding, after writing the above and knowing the bug is at runtime, will the eventual fix address conversions of integral terms to `double` terms that are only known at runtime (i.e., non-literals); or will the fix only address casts of integral literal terms (since those are done at compilation time)? I hope the answer is “yes” for the sake of my ability to comprehend your amazing answer (assuming the fix is *not* to use an intrinsic algorithm). And if so, then one can argue the bug is not even in the runtime but the actual hardware (i.e., a program that expects `ulong` casts to `double`s to be the same as casts of `U128`s to `double`s has a determinism bug caused by a determinism bug in the compiler caused by a determinism bug in the runtime caused by the *actual* bug in the x87 FPU stack).

**Reply by tannergooding** [reply]
It's a bit more complicated than a simple yes/no.

The C# compiler is itself written in C# and it currently depends on the underlying runtime for the `ulong -> double` conversion. This means that C# compilation time behavior matches the runtime behavior.
However, there are also multiple target runtimes namely "full framework" and ".net core" (there is also "mono" and theoretically some 3rd party runtime that might exist) so if we fix it in .NET Core, it will remain broken in full framework and the result you get will be dependent on whether you are using `C:\Program Files (x86)\Microsoft Visual Studio\<Version>\MSBuild\Current\Bin\Roslyn\csc.exe` (which targets full framework) or `C:\Program Files\dotnet\sdk\<version>\Roslyn\bincore\csc.dll` (which targets .net core).

The compiler will, realistically, need to take a fix itself so that the compilation is deterministic on all platforms (x86, x64, ARM32, ARM64; Windows, Linux, OSX, etc; and all variants there-in) and regardless of the runtime it is on (full framework, .net core, mono, etc).

> And if so, then one can argue the bug is not even in the runtime but the actual hardware

There is no bug in the hardware here. The fault lies solely in the algorithm the runtime is using for the conversion: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/codegenxarch.cpp#L6435-L6468

**Reply by tannergooding** [reply]
That is to say, both the runtime and the compiler need to take similar fixes.

The runtime needs it to ensure correctness on future releases and the compiler needs it to ensure correctness regardless of the underlying runtime since only one of the runtimes it runs on will get fixed.

**Reply by zacknewman** [reply]
But isn’t the x87 FPU stack which only has 80-bits of precision non-compliant with IEEE 754 and (per the proof) will not *always* perform binary64 arithmetic correctly since 64 ≠ 80 and 80 < 64*n* for natural numbers *n* > 1?

**Reply by tannergooding** [reply]
The x87 FPU stack isn't really used by anything other than the 32-bit legacy JIT on full framework. .NET Core (and RyuJIT in general) uses SSE/SSE2 and only has to use the x87 FPU stack for where it required by the calling convention on 32-bit machines (namely for returning floating-point values). This has been the case for both 32-bit and 64-bit since .NET Core 2.1 and for 64-bit for quite a while prior to that even extending back to full framework.

Also, while the underlying registers are indeed 80-bits, the x87 FPU stack supports explicitly setting the rounding mode to `32-bits`, `64-bits`, or `80-bits`. When using the x87 FPU stack, .NET has traditionally set this rounding mode to 64-bits which (as per the paper I linked above) avoids the double rounding issues for most code. You are then only left with potential differences if you are doing `32-bit` float operations and not explicitly rounding down to 32-bits after each operation or in cases where the underlying algorithm is wrong/buggy.

**Reply by zacknewman** [reply]
Ah! As is often the case with natural language, it took some time and extended dialogue for the interpretations of both parties to converge to the same thing; although I am sure mine is still “off” since I am a bit out of my element here.

**Reply by tannergooding** [reply]
Happy to answer any further questions you might have. Communicating everything correctly can be difficult 😄
